#Algoritmos - Actividad Guiada 1

Nombre: Rafael Collado Molina<br>
URL: https://github.com/BLueFiReW/Algoritmos-de-Optimizacion/

##Torres de Hanoi con Divide y vencerás

In [2]:
def Torres_Hanoi(N, desde, hasta):
  if N ==1 :
    print("Lleva la ficha " ,desde , " hasta ", hasta )

  else:
    #Torres_Hanoi(N-1, desde, 6-desde-hasta )
    # El 6 sale del 1+2+3 (Torre 1, Torre 2 y Torre 3)
    Torres_Hanoi(N-1, desde, 6-desde-hasta ) #Va de la torre 1 a la 3, entonces 6-1-3 = 2
    print("Lleva la ficha " ,desde , " hasta ", hasta )
    #Torres_Hanoi(N-1,6-desde-hasta, hasta )
    Torres_Hanoi(N-1, 6-desde-hasta  , hasta )


Torres_Hanoi(3, 1 , 3)

# Roja
# Amarilla
# Verde

# -> Torres_Hanoi(2, 1, 2)
# -> -> Torres_Hanoi(1, 1, 3)
# -> -> -> print("Llevar ficha Roja desde la Torre 1 hasta la Torre 3) [Amarilla sobre la Verde   0  la torre 3 Roja) # En este caso es ya N = 1
# -> -> Torres_Hanoi(1, 1, 2)
# -> -> -> print(Lleva ficha Amarilla desde 1 hasta 2). [Verde Amarilla Roja]
# -> -> Torres_Hanoi(1, 3, 2)
# -> -> -> print(Lleva ficha Roja desde 3 hasta 2). [Verde  Roja//Amarilla   0] #Para N = 1
# -> Torres_Hanoi(2, 2, 3)
# -> -> Torres_Hanoi(1, 2, 1)
# -> -> -> print(Lleva ficha Roja desde 2 hasta 1). [Amarilla//Verde//Roja] #Para N = 1
# -> -> Torres_Hanoi(1, 1, 3)
# [Roja 0 Amarilla//Verde]
# [0 0 Roja//Amarilla//Verde]

Lleva la ficha  1  hasta  3
Lleva la ficha  1  hasta  2
Lleva la ficha  3  hasta  2
Lleva la ficha  1  hasta  3
Lleva la ficha  2  hasta  1
Lleva la ficha  2  hasta  3
Lleva la ficha  1  hasta  3


In [3]:
#Sucesión_de_Fibonacci
#https://es.wikipedia.org/wiki/Sucesi%C3%B3n_de_Fibonacci
#Calculo del termino n-simo de la suscesión de Fibonacci
def Fibonacci(N:int):
  if N < 2:
    return 1
  else:
    return Fibonacci(N-1)+Fibonacci(N-2)

Fibonacci(5)

8

##Devolución de cambio por técnica voraz

In [4]:
def cambio_monedas(N, SM):
  SOLUCION = [0]*len(SM)   #SOLUCION = [0,0,0,0,..]
  ValorAcumulado = 0

  for i,valor in enumerate(SM):
    monedas =  (N-ValorAcumulado)//valor
    SOLUCION[i] = monedas
    ValorAcumulado = ValorAcumulado + monedas*valor

    if ValorAcumulado == N:
      return SOLUCION


cambio_monedas(15,[25,10,5,1])

[0, 1, 1, 0]

## N-Reinas por técnica de vueta atrás


In [5]:
def escribe(S):
  n = len(S)
  for x in range(n):
    print("")
    for i in range(n):
      if S[i] == x+1:
        print(" X " , end="")
      else:
        print(" - ", end="")


def es_prometedora(SOLUCION,etapa):
  #print(SOLUCION)
  #Si la solución tiene dos valores iguales no es valida => Dos reinas en la misma fila
  for i in range(etapa+1):
    #print("El valor " + str(SOLUCION[i]) + " está " +  str(SOLUCION.count(SOLUCION[i])) + " veces")
    if SOLUCION.count(SOLUCION[i]) > 1:
      return False

    #Verifica las diagonales
    for j in range(i+1, etapa +1 ):
      #print("Comprobando diagonal de " + str(i) + " y " + str(j))
      if abs(i-j) == abs(SOLUCION[i]-SOLUCION[j]) : return False
  return True



def reinas(N, solucion=[], etapa=0):
  if len(solucion) == 0:
      solucion=[0 for i in range(N)]

  for i in range(1, N+1):
    solucion[etapa] = i

    if es_prometedora(solucion, etapa):
      if etapa == N-1:
        print(solucion)
        #escribe(solucion)
        print()
      else:
        reinas(N, solucion, etapa+1)
    else:
      None

    solucion[etapa] = 0

reinas(8)

[1, 5, 8, 6, 3, 7, 2, 4]

[1, 6, 8, 3, 7, 4, 2, 5]

[1, 7, 4, 6, 8, 2, 5, 3]

[1, 7, 5, 8, 2, 4, 6, 3]

[2, 4, 6, 8, 3, 1, 7, 5]

[2, 5, 7, 1, 3, 8, 6, 4]

[2, 5, 7, 4, 1, 8, 6, 3]

[2, 6, 1, 7, 4, 8, 3, 5]

[2, 6, 8, 3, 1, 4, 7, 5]

[2, 7, 3, 6, 8, 5, 1, 4]

[2, 7, 5, 8, 1, 4, 6, 3]

[2, 8, 6, 1, 3, 5, 7, 4]

[3, 1, 7, 5, 8, 2, 4, 6]

[3, 5, 2, 8, 1, 7, 4, 6]

[3, 5, 2, 8, 6, 4, 7, 1]

[3, 5, 7, 1, 4, 2, 8, 6]

[3, 5, 8, 4, 1, 7, 2, 6]

[3, 6, 2, 5, 8, 1, 7, 4]

[3, 6, 2, 7, 1, 4, 8, 5]

[3, 6, 2, 7, 5, 1, 8, 4]

[3, 6, 4, 1, 8, 5, 7, 2]

[3, 6, 4, 2, 8, 5, 7, 1]

[3, 6, 8, 1, 4, 7, 5, 2]

[3, 6, 8, 1, 5, 7, 2, 4]

[3, 6, 8, 2, 4, 1, 7, 5]

[3, 7, 2, 8, 5, 1, 4, 6]

[3, 7, 2, 8, 6, 4, 1, 5]

[3, 8, 4, 7, 1, 6, 2, 5]

[4, 1, 5, 8, 2, 7, 3, 6]

[4, 1, 5, 8, 6, 3, 7, 2]

[4, 2, 5, 8, 6, 1, 3, 7]

[4, 2, 7, 3, 6, 8, 1, 5]

[4, 2, 7, 3, 6, 8, 5, 1]

[4, 2, 7, 5, 1, 8, 6, 3]

[4, 2, 8, 5, 7, 1, 3, 6]

[4, 2, 8, 6, 1, 3, 5, 7]

[4, 6, 1, 5, 2, 8, 3, 7]

[4, 6, 8, 2, 7, 1, 3, 5]

[4, 6, 8, 3,

## Viaje por el rio. Programación dinámica

In [6]:
TARIFAS = [
[0,5,4,3,999,999,999],
[999,0,999,2,3,999,11],
[999,999, 0,1,999,4,10],
[999,999,999, 0,5,6,9],
[999,999, 999,999,0,999,4],
[999,999, 999,999,999,0,3],
[999,999,999,999,999,999,0]
]



################################################################
def Precios(TARIFAS):
################################################################
  #Total de Nodos
  N = len(TARIFAS[0])

  #Inicialización de la tabla de precios
  PRECIOS = [ [9999]*N for i in [9999]*N]
  RUTA = [ [""]*N for i in [""]*N]

  for i in range(0,N-1):
    RUTA[i][i] = i             #Para ir de i a i se "pasa por i"
    PRECIOS[i][i] = 0          #Para ir de i a i se se paga 0
    for j in range(i+1, N):
      MIN = TARIFAS[i][j]
      RUTA[i][j] = i

      for k in range(i, j):
        if PRECIOS[i][k] + TARIFAS[k][j] < MIN:
            MIN = min(MIN, PRECIOS[i][k] + TARIFAS[k][j] )
            RUTA[i][j] = k          #Anota que para ir de i a j hay que pasar por k
        PRECIOS[i][j] = MIN

  return PRECIOS,RUTA
################################################################


PRECIOS,RUTA = Precios(TARIFAS)
#print(PRECIOS[0][6])

print("PRECIOS")
for i in range(len(TARIFAS)):
  print(PRECIOS[i])

print("\nRUTA")
for i in range(len(TARIFAS)):
  print(RUTA[i])

#Determinar la ruta con Recursividad
def calcular_ruta(RUTA, desde, hasta):
  if desde == hasta:
    #print("Ir a :" + str(desde))
    return ""
  else:
    return str(calcular_ruta( RUTA, desde, RUTA[desde][hasta])) +  \
                ',' + \
                str(RUTA[desde][hasta] \
              )

print("\nLa ruta es:")
calcular_ruta(RUTA, 0,6)

PRECIOS
[0, 5, 4, 3, 8, 8, 11]
[9999, 0, 999, 2, 3, 8, 7]
[9999, 9999, 0, 1, 6, 4, 7]
[9999, 9999, 9999, 0, 5, 6, 9]
[9999, 9999, 9999, 9999, 0, 999, 4]
[9999, 9999, 9999, 9999, 9999, 0, 3]
[9999, 9999, 9999, 9999, 9999, 9999, 9999]

RUTA
[0, 0, 0, 0, 1, 2, 5]
['', 1, 1, 1, 1, 3, 4]
['', '', 2, 2, 3, 2, 5]
['', '', '', 3, 3, 3, 3]
['', '', '', '', 4, 4, 4]
['', '', '', '', '', 5, 5]
['', '', '', '', '', '', '']

La ruta es:


',0,2,5'

<h2 style="color:#d00000;">✅ MEJORA PARA OPTAR AL 10/10: DOS PUNTOS MÁS CERCANOS</h2>
<ul>
  <li>1D: fuerza bruta vs Divide y Vencerás + complejidad + tiempos</li>
  <li>Extensión: 2D y 3D</li>
</ul>

## 1) 1D (lista de números)

### 1.1 Primer intento: Fuerza bruta

- **Idea:** probar todas las parejas y quedarse con la mínima distancia.
- **Complejidad:** hay aproximadamente \( \frac{n(n-1)}{2} \) parejas, por tanto **\(O(n^2)\)** en tiempo y **\(O(1)\)** en memoria.

In [11]:
def closest_1d_bruteforce(xs):
    best = float("inf")
    best_pair = None
    n = len(xs)
    for i in range(n):
        for j in range(i+1, n):
            d = abs(xs[i] - xs[j])
            if d < best:
                best = d
                best_pair = (xs[i], xs[j])
    return best, best_pair

xs = [5,1,9,11]
closest_1d_bruteforce(xs)

(2, (9, 11))

### ¿Se puede mejorar?

Sí. En 1D, si **ordenamos** la lista, el par de puntos más cercano debe estar entre **elementos adyacentes** en el orden.
Esto permite calcular la distancia mínima comparando solo vecinos consecutivos tras ordenar.

In [24]:
def closest_1d_sorted(xs):
    """
    Mejora para 1D: ordenar y comparar solo adyacentes.
    Tiempo: O(n log n) por la ordenación + O(n) del barrido.
    Memoria extra: depende de la implementación de sorted (típicamente O(n)).
    """
    if len(xs) < 2:
        raise ValueError("Se necesitan al menos 2 números.")

    xs_sorted = sorted(xs)
    best = float("inf")
    best_pair = None

    for i in range(len(xs_sorted) - 1):
        d = xs_sorted[i + 1] - xs_sorted[i]  # ya ordenado => no hace falta abs
        if d < best:
            best = d
            best_pair = (xs_sorted[i], xs_sorted[i + 1])

    return best, best_pair

# Ejemplo
xs = [5, 1, 9, 11]
print(closest_1d_sorted(xs))  # (2, (9, 11))

(2, (9, 11))


### 1.2 Segundo intento: Divide y Vencerás

Para que sea **Divide y Vencerás (DyV) “de libro”**, lo más limpio es usar **MergeSort** (que sigue DyV).  
Una vez obtenida la lista ordenada, hacemos un **barrido lineal** comparando solo **elementos adyacentes**.

Esto respeta la definición de DyV del material: **dividir** en subproblemas, **resolver** recursivamente y **combinar**.

**Complejidad:**

- **MergeSort:** cumple una recurrencia del tipo  
  \[
  T(n)=2T\left(\frac{n}{2}\right)+O(n)\ \Rightarrow\ O(n\log n)
  \]
- **Barrido final:** \(O(n)\) (no cambia el orden asintótico).

Por tanto, el algoritmo completo es **\(O(n\log n)\)**.

In [22]:
def merge(left, right):
    """Fusiona dos listas ORDENADAS en una sola lista ordenada. O(n)."""
    i = j = 0
    out = []
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            out.append(left[i])
            i += 1
        else:
            out.append(right[j])
            j += 1
    # añade lo que quede
    out.extend(left[i:])
    out.extend(right[j:])
    return out

def mergesort(xs):
    """Ordena xs con MergeSort (Divide y Vencerás). O(n log n)."""
    if len(xs) <= 1:          # caso base
        return xs[:]          # copia (por seguridad)
    mid = len(xs) // 2        # dividir
    left = mergesort(xs[:mid])   # conquistar
    right = mergesort(xs[mid:])  # conquistar
    return merge(left, right)    # combinar

def closest_1d_divide_conquer(xs):
    xs_sorted = mergesort(xs)
    best = float("inf")
    best_pair = None
    for i in range(len(xs_sorted) - 1):
        d = xs_sorted[i + 1] - xs_sorted[i]  # ya están ordenados => no hace falta abs
        if d < best:
            best = d
            best_pair = (xs_sorted[i], xs_sorted[i + 1])
    return best, best_pair

xs = [5, 1, 9, 11]
print(closest_1d_divide_conquer(xs))  # (2, (9, 11))

(2, (9, 11))


### ¿Se puede mejorar más?

Esto **no aparece en los apuntes**; es **conocimiento general**.  
Si el algoritmo se basa únicamente en **comparaciones**, la necesidad de **ordenar** impone un **límite inferior** de aproximadamente  
\[
\Omega(n \log n)
\]
por lo que no se puede mejorar asintóticamente en el peor caso.

Solo se podría bajar de ahí si se aprovecha **estructura adicional** del problema, por ejemplo:
- números enteros en un rango pequeño \(\rightarrow\) ordenación por **buckets/counting** (sin comparaciones).

## 2) Extensión a 2D: puntos \((x, y)\)

### 2.1 Fuerza bruta

- **Idea:** mismo patrón que en 1D: comparar **todas las parejas** de puntos.
- **Distancia:** euclídea  
  \[
  d(p,q)=\sqrt{(x_p-x_q)^2+(y_p-y_q)^2}
  \]
  (en implementación suele usarse la distancia **al cuadrado** para evitar la raíz).

In [23]:
import math

def dist2(p, q):
    # distancia euclídea al cuadrado en 2D
    return (p[0] - q[0])**2 + (p[1] - q[1])**2

def closest_2d_bruteforce(pts):
    if len(pts) < 2:
        raise ValueError("Se necesitan al menos 2 puntos (x, y).")

    best_d2 = float("inf")
    best_pair = None

    n = len(pts)
    for i in range(n):
        for j in range(i + 1, n):
            d2 = dist2(pts[i], pts[j])
            if d2 < best_d2:
                best_d2 = d2
                best_pair = (pts[i], pts[j])

    # devuelve distancia REAL (no al cuadrado) y la pareja
    return math.sqrt(best_d2), best_pair

pts2d = [(5, 0), (1, 0), (9, 0), (11, 0)]
print(closest_2d_bruteforce(pts2d))  # (2.0, ((9,0),(11,0))) o similar

(2.0, ((9, 0), (11, 0)))


Complejidad: O(n²).

## 3) Extensión a 3D \((x, y, z)\)

### 3.1 Fuerza bruta

- **Complejidad:** \(O(n^2)\) en tiempo.
- **Idea:** comparar **todas las parejas** de puntos.
- **Implementación:** usar la **distancia al cuadrado** para evitar `sqrt` (es más rápido y no cambia qué par minimiza la distancia):
  \[
  d^2(p,q)=(x_p-x_q)^2+(y_p-y_q)^2+(z_p-z_q)^2
  \]

In [21]:
import math
from typing import Iterable, List, Tuple, Optional

Point3D = Tuple[float, float, float]

def dist2_3d(p: Point3D, q: Point3D) -> float:
    """Distancia euclídea al cuadrado en 3D."""
    dx = p[0] - q[0]
    dy = p[1] - q[1]
    dz = p[2] - q[2]
    return dx*dx + dy*dy + dz*dz

def closest_3d_bruteforce(points: List[Point3D]) -> Tuple[float, Tuple[Point3D, Point3D]]:
    """
    Devuelve (distancia_minima, (p, q)) para el par más cercano en 3D.
    Complejidad: O(n^2) tiempo, O(1) memoria extra.
    """
    n = len(points)
    if n < 2:
        raise ValueError("Se necesitan al menos 2 puntos.")

    best_d2 = float("inf")
    best_pair: Optional[Tuple[Point3D, Point3D]] = None

    for i in range(n):
        pi = points[i]
        for j in range(i + 1, n):
            d2 = dist2_3d(pi, points[j])
            if d2 < best_d2:
                best_d2 = d2
                best_pair = (pi, points[j])

    # best_pair no puede ser None porque n>=2
    return math.sqrt(best_d2), best_pair  # distancia real

# -----------------------
# Tests rápidos
# -----------------------
if __name__ == "__main__":
    pts = [(0, 0, 0), (10, 0, 0), (1, 1, 1), (2, 2, 2)]
    d, (p, q) = closest_3d_bruteforce(pts)
    print("dist =", d, "pair =", (p, q))
    # El par más cercano debería ser (1,1,1)-(2,2,2) con dist sqrt(3)≈1.732...

dist = 1.7320508075688772 pair = ((0, 0, 0), (1, 1, 1))


## 4) “Plus” que suele dar puntos: experimento de tiempos + mini-estadística

Como el propio AG1 te pide **generar datos aleatorios**, añade un experimento comparando **fuerza bruta** vs **Divide y Vencerás (1D)**:

- Tamaños: \(n = 200, 500, 1000, 2000, \dots\)
- **10 repeticiones** por cada \(n\)
- Para cada método, reporta **media** y **desviación típica** del tiempo de ejecución

In [25]:
import time
import random
import statistics

def benchmark_1d(ns=(200, 500, 1000, 2000), repeats=10, seed=12345, value_range=10_000_000,
                 brute_fn=None, fast_fn=None, check_equal=True):
    """
    Mide tiempos (ms) y calcula media + desviación típica.
    brute_fn: función O(n^2)
    fast_fn : función mejorada (O(n log n))
    """
    if brute_fn is None:
        brute_fn = globals().get("closest_1d_bruteforce")
    if fast_fn is None:
        # cambia el nombre aquí si tu mejora se llama distinto
        fast_fn = globals().get("closest_1d_divide_conquer") or globals().get("closest_1d_sorted")

    if brute_fn is None or fast_fn is None:
        raise NameError("No encuentro closest_1d_bruteforce y/o closest_1d_divide_conquer (o closest_1d_sorted) en el notebook.")

    results = []
    for n in ns:
        t_brute = []
        t_fast = []

        for r in range(repeats):
            rng = random.Random(seed + 100000 * n + r)
            xs = [rng.randrange(value_range) for _ in range(n)]

            t0 = time.perf_counter()
            d_b, _ = brute_fn(xs)
            t1 = time.perf_counter()
            t_brute.append((t1 - t0) * 1000.0)

            t0 = time.perf_counter()
            d_f, _ = fast_fn(xs)
            t1 = time.perf_counter()
            t_fast.append((t1 - t0) * 1000.0)

            if check_equal and d_b != d_f:
                raise AssertionError(f"Distancias distintas (n={n}, rep={r}): brute={d_b}, fast={d_f}")

        results.append({
            "n": n,
            "brute_mean_ms": statistics.mean(t_brute),
            "brute_std_ms": statistics.pstdev(t_brute),
            "fast_mean_ms": statistics.mean(t_fast),
            "fast_std_ms": statistics.pstdev(t_fast),
        })

    return results

def print_results_table(results):
    print(f"{'n':>6} | {'brute (ms)':>20} | {'mejora (ms)':>20}")
    print("-" * 56)
    for r in results:
        print(f"{r['n']:>6} | "
              f"{r['brute_mean_ms']:>8.2f} ± {r['brute_std_ms']:<8.2f} | "
              f"{r['fast_mean_ms']:>8.2f} ± {r['fast_std_ms']:<8.2f}")

# Ejecución
res = benchmark_1d(ns=(200, 500, 1000, 2000), repeats=10)
print_results_table(res)

     n |           brute (ms) |          mejora (ms)
--------------------------------------------------------
   200 |     1.77 ± 0.50     |     0.38 ± 0.17    
   500 |    12.25 ± 2.95     |     1.00 ± 0.31    
  1000 |    72.38 ± 5.10     |     2.78 ± 0.43    
  2000 |   296.67 ± 12.63    |     6.86 ± 1.38    


### Resultados experimentales (tiempos y mini-estadística)

Se han comparado dos enfoques en 1D:
- `closest_1d_bruteforce`: fuerza bruta (todas las parejas) → complejidad **O(n²)**.
- `closest_1d_divide_conquer`: MergeSort (Divide y Vencerás) + comparación de adyacentes → **O(n log n)**.

Tabla de tiempos (10 repeticiones por tamaño, media ± desviación típica):

 n | brute (ms) | mejora (ms)
---|------------|------------
200  | 1.77 ± 0.50   | 0.38 ± 0.17
500  | 12.25 ± 2.95  | 1.00 ± 0.31
1000 | 72.38 ± 5.10  | 2.78 ± 0.43
2000 | 296.67 ± 12.63| 6.86 ± 1.38

**Interpretación:**
- Al doblar `n`, la fuerza bruta crece aproximadamente ×4 (comportamiento típico de **O(n²)**).
- En cambio, la mejora crece mucho más lentamente (≈ ×2–×2.5 al doblar `n`), coherente con **O(n log n)**.
- El beneficio de la mejora aumenta con el tamaño del problema: a `n=2000` es ~**43×** más rápida que fuerza bruta.